### PR TITLE
New detector `FindOverridableMethodCall`

### PR DESCRIPTION
**Pull request by baloghadamsoftware**
This new detector detects invocation of overridable method in constructors
(`MC_OVERRIDABLE_METHOD_CALL_IN_CONSTRUCTOR`) and clone() method
(`MC_OVERRIDABLE_METHOD_CALL_IN_CLONE`), according to SEI CERT rules
[MET05-J. Ensure that constructors do not call overridable methods](https://wiki.sei.cmu.edu/confluence/display/java/MET05-J.+Ensure+that+constructors+do+not+call+overridable+methods)
and [MET06-J. Do not invoke overridable methods in clone()](https://wiki.sei.cmu.edu/confluence/pages/viewpage.action?pageId=88487921)

----

Make sure these boxes are checked before submitting your PR -- thank you!

- [X] Added an entry into `CHANGELOG.md` if you have changed SpotBugs code

### CODE REVIEW

**Comment by baloghadamsoftware** [comment]
This is the basic version. The plan is to also support indirect calls to find more true positives.

File: spotbugs/etc/findbugs.xml [hunk 1]
```diff
@@ -1254,4 +1256,6 @@
           <BugPattern abbrev="NP" type="NP_METHOD_PARAMETER_TIGHTENS_ANNOTATION" category="STYLE" />
           <BugPattern abbrev="NP" type="NP_METHOD_PARAMETER_RELAXING_ANNOTATION" category="STYLE" deprecated="true" />
           <BugPattern abbrev="EOS" type="EOS_BAD_END_OF_STREAM_CHECK" category="CORRECTNESS" />
+          <BugPattern abbrev="MC" type="MC_OVERRIDABLE_METHOD_CALL_IN_CONSTRUCTOR" category="BAD_PRACTICE" />
+          <BugPattern abbrev="MC" type="MC_OVERRIDABLE_METHOD_CALL_IN_CLONE" category="BAD_PRACTICE" />
```
Comment (KengoTODA) [inline spotbugs/etc/findbugs.xml:1260]: I think `BAD_PRACTICE` is acceptable, but [this page](https://wiki.sei.cmu.edu/confluence/display/java/MET05-J.+Ensure+that+constructors+do+not+call+overridable+methods) pointed that code with this problem may lead to vulnerability then the [`MALICIOUS_CODE` cateogry](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#malicious-code-vulnerability-malicious-code) is also our option.

**Review by KengoTODA — APPROVED** [review]
LGTM, there is one thing to discuss for providing an intuitive category to users.

File: spotbugs/etc/findbugs.xml (hunk 1 above)
Comment (baloghadamsoftware) [reply spotbugs/etc/findbugs.xml:1260]: I absolutely agree with you, this bug could land in any of these categories. Now I feel that MALICIOUS_CODE is maybe better than BAD_PRACTICE. Should I change it? Or maybe should we wait for others to vote, e.g. @ThrawnCA?

**Comment by baloghadamsoftware** [comment]
Our findings on Jenkins, [OpenGrok, SpotBugs itself, NanoHTTPD, Ttorrent, Bt and MATSim-Libs](https://codechecker-demo.eastus.cloudapp.azure.com/java-static-analysis/reports?items-per-page=50&run=SpotBugs%20on%20MATSim-Libs%20%28user%3A%20spotbugs,%20branch%3A%20master%29&run=SpotBugs%20on%20SpotBugs%20%28user%3A%20spotbugs,%20branch%3A%20master%29&run=SpotBugs%20on%20Jenkins%20%28user%3A%20spotbugs,%20branch%3A%20master%29&run=SpotBugs%20on%20Bt%20%28user%3A%20spotbugs,%20branch%3A%20master%29&run=SpotBugs%20on%20OpenGrok%20%28user%3A%20spotbugs,%20branch%3A%20master%29&run=SpotBugs%20on%20Ttorrent%20%28user%3A%20spotbugs,%20branch%3A%20master%29&run=SpotBugs%20on%20NanoHTTPD%20%28user%3A%20spotbugs,%20branch%3A%20master%29&newcheck=SpotBugs%20on%20MATSim-Libs%20%28user%3A%20baloghadamsoftware,%20branch%3A%20DangerousOverridableMethodCall%29&newcheck=SpotBugs%20on%20Jenkins%20%28user%3A%20baloghadamsoftware,%20branch%3A%20DangerousOverridableMethodCall%29&newcheck=SpotBugs%20on%20SpotBugs%20%28user%3A%20baloghadamsoftware,%20branch%3A%20DangerousOverridableMethodCall%29&newcheck=SpotBugs%20on%20Ttorrent%20%28user%3A%20baloghadamsoftware,%20branch%3A%20DangerousOverridableMethodCall%29&newcheck=SpotBugs%20on%20Bt%20%28user%3A%20baloghadamsoftware,%20branch%3A%20DangerousOverridableMethodCall%29&newcheck=SpotBugs%20on%20OpenGrok%20%28user%3A%20baloghadamsoftware,%20branch%3A%20DangerousOverridableMethodCall%29&newcheck=SpotBugs%20on%20NanoHTTPD%20%28user%3A%20baloghadamsoftware,%20branch%3A%20DangerousOverridableMethodCall%29).

**Comment by KengoTODA** [comment]
What is the difference between #1716 and this PR? Is it OK to close this PR to focus on #1716? 🤔

**Comment by baloghadamsoftware** [comment]
> 
> 
> What is the difference between #1716 and this PR? Is it OK to close this PR to focus on #1716? 🤔

This is the base PR which only recognizes direct calls to overridable methods from constructors and `clone()` methods. I tried to work incrementally to make reviews easier. #1716 uses a simple graph algorithm to also detect cases where a constructor or `clone()` calls a non-overridable which may also call another non-overridable and so on, but eventually one of them calling an overridable. We get a few additional findings with that one. Feel free to close this one and review the more complex PR.

**Comment by KengoTODA** [comment]
> Feel free to close this one and review the more complex PR.

I will do so.

**Pull request closed without merging**